### PR TITLE
[Bugfix][Quantization] Register Quark per-block FP8 scales as weight_scale

### PR DESCRIPTION
**Pull request by jimmy-adams**
## Summary
- Register Quark per-block FP8 linear scales as `weight_scale`, matching what Quark exports, instead of `weight_scale_inv` (DeepSeek-V3 `Fp8LinearMethod` convention).
- Drop the DeepSeek-V4 AMD mapper rewrite of non-expert `.weight_scale` → `.weight_scale_inv`. Loaders already alias to `_inv` only when that parameter is registered (`Fp8LinearMethod` / MXFP4 override).
- Read block scales through `get_fp8_block_weight_scale()` so DSV4 ROCm preshuffle and MLA `wo_a` dequant work with Quark layers.

This is the quant-method fix discussed on #54556 / #52263 / #47972, rather than another model-specific remap.

## Test plan
- [x] `pytest tests/quantization/test_quark.py::test_quark_w8a8_fp8_per_block_registers_weight_scale`
- [x] Load `amd/GLM-5.3-Quark-MXFP4` (or equivalent) without `KeyError` on `.weight_scale` — GSM8K 0.9158, see [this comment](https://github.com/vllm-project/vllm/pull/54770#issuecomment-5506943426)
- [x] Load DeepSeek-V4 AMD Quark NVFP4 (`QuarkW8A8Fp8PerBlock`) and MXFP4 (`Fp8LinearMethod` + `_resolve_param_name`) — NVFP4 verified (GSM8K 0.9507, see [this comment](https://github.com/vllm-project/vllm/pull/54770#issuecomment-5506943426)); MXFP4 still pending
- [x] Confirm native DeepSeek `.scale` → `weight_scale_inv` path is unchanged

### CODE REVIEW

File: vllm/model_executor/layers/quantization/quark/schemes/quark_w8a8_fp8.py [hunk 1]
```diff
@@ -230,6 +230,9 @@ def get_min_capability(cls) -> int:
         return QuarkW8A8Fp8.get_min_capability()
 
     def process_weights_after_loading(self, layer) -> None:
+        # Quark exports the dequant multiplier as ``weight_scale`` (the same
+        # numerical convention as DeepSeek's ``weight_scale_inv``). Kernels
+        # multiply ``weight * scale``; do not invert here.
```
Comment (fxmarty-amd) [inline vllm/model_executor/layers/quantization/quark/schemes/quark_w8a8_fp8.py:235]: Does Quark fp8 per-block `weight_scale` convention matches the rest of quark exported `weight_scale`?

If so, it sounds good to me. We should just have a single convention for quark exported models.

File: vllm/model_executor/layers/quantization/quark/schemes/quark_w8a8_fp8.py [hunk 2]
```diff
@@ -280,8 +283,10 @@ def create_weights(
             weight_loader,
             scale_dtype=scale_dtype,
         )
-        # DeepSeek V4 weight mappers route checkpoint ".scale" tensors here.
-        layer.register_parameter("weight_scale_inv", weight_scale)
+        # Match Quark-exported checkpoints (``.weight_scale``). DeepSeek-V3
+        # native block-FP8 still registers ``weight_scale_inv`` on
+        # ``Fp8LinearMethod``; that path is unchanged.
+        layer.register_parameter("weight_scale", weight_scale)
```
Comment (fxmarty-amd) [inline vllm/model_executor/layers/quantization/quark/schemes/quark_w8a8_fp8.py]:
```suggestion
        # Match Quark-exported checkpoints (``.weight_scale``).
        # ``Fp8LinearMethod.create_weights`` registers weight_scale_inv, which has the
        # same semantic meaning as Quark's `weight_scale`.
        layer.register_parameter("weight_scale", weight_scale)
```

if this is correct

File: vllm/model_executor/layers/quantization/utils/fp8_utils.py [hunk 3]
```diff
@@ -45,6 +45,24 @@ def is_fp8(x: torch.dtype | torch.Tensor) -> bool:
     return x == torch.float8_e4m3fn or x == torch.float8_e4m3fnuz
 
 
+def get_fp8_block_weight_scale(layer: torch.nn.Module) -> torch.Tensor | None:
```
Comment (fxmarty-amd) [inline vllm/model_executor/layers/quantization/utils/fp8_utils.py:53]: Can we have this handle explicitly `isinstance(layer.quant_method, QuarkW8A8Fp8PerBlock)`, and return `"weight_scale_inv"` in any other case?

File: vllm/models/deepseek_v4/amd/model.py [hunk 4]
```diff
@@ -922,20 +925,17 @@ def _make_deepseek_v4_weights_mapper(
...
         #  - DeepSeek native ``.scale``: expert scales -> ``.weight_scale``,
         #    everything else -> ``.weight_scale_inv``.
-        #  - AMD-Quark ``.weight_scale``: linear/attn scales ->
-        #    ``.weight_scale_inv``. Expert and shared-expert
-        #    ``w{1,2,3}.weight_scale`` are left untouched (consumed as-is by
-        #    the MXFP4 expert loader, which produces ``w{13,2}_weight_scale``);
         scale_regex = {
             re.compile(r"(\.experts\.\d+\.w[123])\.scale$"): r"\1.weight_scale",
             re.compile(r"\.scale$"): ".weight_scale_inv",
-            re.compile(r"(?<!\.w[123])\.weight_scale$"): ".weight_scale_inv",
```
Comment (fxmarty-amd) [inline vllm/models/deepseek_v4/amd/model.py:938]: Might be safer to remove only if if `isinstance(quant_config, QuarkConfig)`? As we are not validating this change beyond QuarkConfig.

File: vllm/models/deepseek_v4/amd/mtp.py [hunk 5]
```diff
@@ -335,11 +335,10 @@ def _find_mtp_layer_idx(name: str) -> int:
...
-            # while MXFP4 experts register ``.weight_scale``. Auto-detect:
-            # rename to ``_inv`` only when that variant exists and the plain
-            # one does not.
+            # Quark checkpoints and QuarkW8A8Fp8PerBlock use ``.weight_scale``.
+            # Native block-FP8 layers register ``.weight_scale_inv``. Rename
+            # to ``_inv`` only when that variant exists and the plain one
+            # does not.
             if name.endswith(".weight_scale") and name not in params_dict:
                 inv = name.removesuffix(".weight_scale") + ".weight_scale_inv"
                 if inv in params_dict:
```
Comment (fxmarty-amd) [inline vllm/models/deepseek_v4/amd/mtp.py:344]: nit (ignore for this PR): cc @ColinZ22 

This was added in https://github.com/vllm-project/vllm/pull/48044.

Given that https://github.com/vllm-project/vllm/pull/47972 was merged, I don't think we need to keep this?

https://github.com/vllm-project/vllm/pull/48044/changes#r3562411245

> The quark schema doesn't map to the flat kwargs Fp8Config expects, and the DSV4 attn are blockwise FP8, so we translate to reuse the existing blockwise FP8 + MXFP4-MoE paths instead of duplicating our own.

I don't think the good place for this kind of logic is in a specific model (ds-v4), given that quark-produced models of different model typesmay use FP8 per-block as well?

Should this be cleaned up?

File: vllm/models/deepseek_v4/amd/model.py [hunk 6]
```diff
@@ -774,6 +774,9 @@ def load_weights(self, weights: Iterable[tuple[str, torch.Tensor]]) -> set[str]:
         loaded_params: set[str] = set()
 
         def _resolve_param_name(name: str) -> str:
+            # Fp8LinearMethod registers block scales as ``weight_scale_inv``.
+            # Quark checkpoints / QuarkW8A8Fp8PerBlock use ``weight_scale``.
+            # Alias only when the registered param is the ``_inv`` name.
             inv_name = f"{name}_inv"
             if name not in params_dict and inv_name in params_dict:
                 return inv_name
```
Comment (fxmarty-amd) [inline vllm/models/deepseek_v4/amd/model.py:782]: Same comment about this that was added in https://github.com/vllm-project/vllm/pull/47972

File: vllm/model_executor/layers/quantization/quark/schemes/quark_w8a8_fp8.py (hunk 1 above)
Comment (jimmy-adams) [reply vllm/model_executor/layers/quantization/quark/schemes/quark_w8a8_fp8.py:235]: Yes, Quark regulates the export format and name it as the weight_scale.

File: vllm/models/deepseek_v4/amd/model.py (hunk 4 above)
Comment (jimmy-adams) [reply vllm/models/deepseek_v4/amd/model.py:938]: That rule only matched keys already spelled `.weight_scale` (Quark). Native DeepSeek spells them `.scale`, handled by the rule that stays, and the NVIDIA/XPU mappers never had this rewrite — so dropping it converges AMD with them. Mixed checkpoints are still covered by `_resolve_param_name`.

File: vllm/model_executor/layers/quantization/utils/fp8_utils.py (hunk 3 above)
Comment (jimmy-adams) [reply vllm/model_executor/layers/quantization/utils/fp8_utils.py:53]: Hi @fxmarty-amd Done. One correction: `QuarkW8A8Fp8PerBlock` is stored in `layer.scheme`, while `layer.quant_method` is `QuarkLinearMethod`. The helper now checks `layer.scheme` and returns `weight_scale_inv` otherwise. Tests pass.

File: vllm/models/deepseek_v4/amd/mtp.py (hunk 5 above)
Comment (jimmy-adams) [reply vllm/models/deepseek_v4/amd/mtp.py:344]: It is still needed for Quark MXFP4: that path is translated to `Fp8Config`, so the layer registers `weight_scale_inv` while the checkpoint exports `weight_scale`. Agreed this should be moved out of the DSV4 loader in a follow-up.

**Comment by jimmy-adams** [comment]
### End-to-end validation on real Quark checkpoints

Both runs are at this PR's head `c17d120d7`, on ROCm / AMD MI350/MI355
(`gfx950`).

#### 1. GLM-5.3 MXFP4 + attention FP8 — `QuarkW8A8Fp8PerBlock`

Covers the "Load `amd/GLM-5.3-Quark-MXFP4` without `KeyError` on
`.weight_scale`" item.
Model: [`amd/GLM-5.3-Quark-MXFP4-AttnFP8`](https://huggingface.co/amd/GLM-5.3-Quark-MXFP4-AttnFP8)
(`GlmMoeDsaForCausalLM`), TP=4.

`VLLM_USE_V2_MODEL_RUNNER=1` is required here for reasons unrelated to this PR:
GLM-5.3 on ROCm is currently routed to the V1 model runner by #53155, which
corrupts generations for this architecture with or without these changes. That
regression is tracked in #54924, with a fix pending in #54925; until it lands,
the override is needed to measure this PR at all.

```bash
MODEL=amd/GLM-5.3-Quark-MXFP4-AttnFP8

VLLM_USE_V2_MODEL_RUNNER=1 \
VLLM_ROCM_USE_AITER=1 \
VLLM_ROCM_USE_AITER_FP8BMM=0 \
VLLM_ROCM_USE_AITER_FP4BMM=0 \
python -m lm_eval --model vllm \
  --model_args "pretrained=$MODEL,tensor_parallel_size=4,dtype=auto,\
quantization=quark,max_model_len=32768,trust_remote_code=True,max_num_seqs=32" \
  --tasks gsm8k \
  --batch_size auto
```

GSM8K 5-shot, 1319 docs:

| Build | flexible-extract | strict-match |
| --- | --- | --- |
| this PR (`c17d120d7`) + `VLLM_USE_V2_MODEL_RUNNER=1` | **0.9158** | **0.9143** |

Loads with no `KeyError` on `.weight_scale`, and accuracy is slightly above the
0.9128 the model card reports.

#### 2. DeepSeek-V4-Flash NVFP4 — Quark NVFP4 path

Covers the NVFP4 half of the "Load DeepSeek-V4 AMD Quark NVFP4" item.
Model: `amd/DeepSeek-V4-Flash-NVFP4` (`DeepseekV4ForCausalLM`,
`quant_method: quark`, fp4 `per_group` / `group_size: 16`), TP=8.

No runner override here: `DeepseekV4ForCausalLM` was already in
`ROCM_DEFAULT_MRV1_ARCHITECTURES` before #53155, so this serves on MRV1 by
default and is unaffected — which also shows the corruption in #54924 is
specific to `GlmMoeDsaForCausalLM` rather than MRV1 in general.

```bash
VLLM_ROCM_USE_AITER=1 \
VLLM_ROCM_USE_AITER_MOE=1 \
vllm serve amd/DeepSeek-V4-Flash-NVFP4 \
  --host localhost --port 8001 \
  --dtype auto \
  --kv-cache-dtype fp8 \
  --tensor-parallel-size 8 \
  --max-num-seqs 512 \
  --max-num-batched-tokens 8192 \
  --distributed-executor-backend mp \
  --trust-remote-code \
  --gpu-memory-utilization 0.9 \
  --tokenizer-mode deepseek_v4 \
  --reasoning-parser deepseek_v4 \
  --tool-call-parser deepseek_v4 \
  --enable-auto-tool-choice \
  --compilation-config '{"mode": 3, "cudagraph_mode": "FULL_DECODE_ONLY"}'

lm_eval --model local-completions \
  --model_args model=amd/DeepSeek-V4-Flash-NVFP4,\
tokenizer=amd/DeepSeek-V4-Flash-NVFP4,\
base_url=http://127.0.0.1:8001/v1/completions,\
num_concurrent=32,max_retries=10,max_gen_toks=2048,timeout=60000 \
  --batch_size auto \
  --tasks gsm8k \
  --num_fewshot 8
```

GSM8K 8-shot, 1319 docs:

| Build | flexible-extract | strict-match |
| --- | --- | --- |
| this PR (`c17d120d7`) | **0.9507** | **0.9515** |

The Quark NVFP4 checkpoint loads and serves cleanly with the `.weight_scale`
registration and `get_fp8_block_weight_scale()` changes in this PR.

**Review by fxmarty-amd — APPROVED** [review]
LGTM.

Deepseek v4 should be covered in LM-Eval CI @jimmy-adams ? Let's check that

**Comment by jimmy-adams** [comment]
> LGTM.
> 
> Deepseek v4 should be covered in LM-Eval CI @jimmy-adams ? Let's check that

Hi @fxmarty-amd, The DS-V4 test result is also attached. Thanks for reminding.

**Comment by jimmy-adams** [comment]
### CI failures here are Buildkite checkout errors, not this PR

Every failed job in build [86845](https://buildkite.com/vllm/ci/builds/86845)
died during "Preparing working directory" and never executed a test:

```
$ git fetch -v --prune -- origin refs/pull/54770/head c17d120d7b315331d105f193d0f4b9dc7bab5e00
fatal: could not read Username for 'https://github.com': terminal prompts disabled
fatal: expected flush after ref listing
🚨 Error: cloning git repository: exit status 128
```

Evidence that this is agent-side rather than PR-specific:

- 20/20 failed jobs end in `exit status 128` after exhausting checkout retries.
  None contains an assertion, `FAILED` line, OOM, or timeout.
- The failures span 10 distinct agents in 6 host groups
  (`h200-ci-2-9/21/23/24`, `h200-ci-3-13`, `h200-ci-4-23/32`, `h200-ci-5-46`,
  `h200-ci-6-30/31`), so it is not one unhealthy agent.
- Postmerge builds on `main` hit the same error in the same window: build 86841
  (`1c26e57d`) had five `nvidia-h200-mig-35gb-*` jobs fail at 09:33:33-09:33:50Z,
  and build 86839 (`584e8f0d`) failed `v1-core-plus-kv-plus-metrics` at 09:40:02Z.
- It looks intermittent rather than a static misconfiguration: in 5 of the 20
  jobs the initial clone completed (`Receiving objects: 100%`) and only the
  subsequent `refs/pull/54770/head` fetch failed seconds later on the same agent.

Worth noting for review: `quantization-shard-1/2/3/4` and `model-executor` are
among the jobs that never ran, so the suites that would actually exercise these
Quark FP8 changes have not been validated yet.

Could a maintainer retrigger build 86845? Happy to rebase or push an empty
commit instead if that is the easier way to get a fresh run.

**Comment by jimmy-adams** [comment]
@AndreasKaratzas  @tjtanaa  @hongxiayang  @dllehr-amd  @shen-shanshan  Can you help reinitiate the CI? This PR had a failure of vllm machine initiation like below.

git fetch -v --prune -- origin refs/pull/54770/head c17d120d7...
fatal: could not read Username for 'https://github.com': terminal prompts disabled

**Review by shen-shanshan — APPROVED** [review]
